reps/2022-10-11-serve-java-http-ingress.md

```diff
@@ -0,0 +1,167 @@
+## Summary
+### General Motivation
+We want to call various methods of java deployment through http proxy. Like the fastapi ingress in python serve.
+### Should this change be within `ray` or outside?
+main `ray` project. Changes are made to Ray Serve module.
+
+## Stewardship
+### Required Reviewers
+@simon-mo
+### Shepherd of the Proposal (should be a senior committer)
+@simon-mo
+
+## Design and Architecture
+### user case
+1. define the model with JAX-RS
+```java
+@Path("user")
+public class UserRestService {
+    @GET
+    @Path("helloWorld")
+    @Consumes({"application/json"})
+    @Produces({"application/json"})
+    public String helloWorld(String name) {
+        return "hello world, " + name;
+    }
+
+    @GET
+    @Path("paramPathTest/{name}")
+    @Consumes({"application/json"})
+    @Produces({"application/json"})
+    public String paramPathTest(@PathParam("name")String name) {
+        return "paramPathTest, " + name;
+    }
+}
+```
+2. create deployment
+```java
+Deployment deployment =
+        Serve.deployment()
+            .setName("deploymentName")
+            .setDeploymentDef(UserRestService.class.getName())
+            .ingress("jax-rs")
+            .setNumReplicas(1)
+            .create();
+deployment.deploy(true);
+```
+3. calling deployments via HTTP
+The URI is determined by the deployment name and JAX-RS @PATH annotations.
+```java
+curl http://127.0.0.1:8000/deploymentName/user/helloWorld?name=test
+curl http://127.0.0.1:8000/deploymentName/user/paramPathTest/test
+```
+### http ingress
+#### ingress api annotation: JAX-RS
+In java application development, restful api is generally implemented through two sets of annotations, spring web or jax-rs.
+
+JAX-RS is a specification for implementing REST web services in Java, currently defined by the JSR-370. JAX-RS is just an API specification and  has been implemented by many components: jersey,resteasy,etc.
+
+Spring-web contains a lot of features that we don't need to use: IOC, DI, etc.The parsing of spring-web annotations depends on the spring framework.
+
+In order not to import too many unnecessary dependencies to the user's `Callable`. We choose JAX-RS to support http ingress.
+
+#### annotation parser: JERSEY
+Jersey RESTful Web Services 2.x framework is open source, production quality, framework for developing RESTful Web Services in Java that provides support for JAX-RS APIs and serves as a JAX-RS (JSR 311 & JSR 339 & JSR 370) Reference Implementation .
```

Comment:
Add link of JERSEY official webpage.

Reply:
Resolved.
For more information, please click this link: https://eclipse-ee4j.github.io/jersey/